pkg/provider/consulcatalog/consul_catalog.go

```diff
@@ -41,12 +44,54 @@ type itemData struct {
 	ExtraConf  configuration
 }
 
-// Provider holds configurations of the provider.
-type Provider struct {
+// ProviderBuilder is responsible to construct namespaced instances of the Consul Catalog provider.
+type ProviderBuilder struct {
+	Configuration `export:"true"`
+
+	// Deprecated: use Namespaces option instead.
+	Namespace  string   `description:"Sets the namespace used to discover services (Consul Enterprise only)." json:"namespace,omitempty" toml:"namespace,omitempty" yaml:"namespace,omitempty"`
+	Namespaces []string `description:"Sets the namespaces used to discover services (Consul Enterprise only)." json:"namespaces,omitempty" toml:"namespaces,omitempty" yaml:"namespaces,omitempty"`
+}
+
+// SetDefaults sets the default values.
+func (p *ProviderBuilder) SetDefaults() {
+	p.Configuration.SetDefaults()
+}
+
+// BuildProviders builds Consul Catalog provider instances for the given namespaces configuration.
+func (p *ProviderBuilder) BuildProviders() []*Provider {
+	if p.Namespace != "" {
```

Comment:
```suggestion
    // we can warn about that, because we've already made sure before that
    // Namespace and Namespaces are mutually exclusive
	if p.Namespace != "" {
```